File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True :
    s = input()
    stack = []

    if s == "." : #종료조건
        break

    for i in s :
        if i == '[' or i == '(' : #여는건 넣기
            stack.append(i)
        elif i == ']' :# [이 있는 경우엔 짝이 맞으니 지우기
            if len(stack) != 0 and stack[-1] == '[' :
                stack.pop()
            else : # 아닐 경우엔 스택 비워지지만 않게 하기
                stack.append(1)
                break
        elif i == ')' : # 위와 마찬가지 반복
            if len(stack) != 0 and stack[-1] == '(' :
                stack.pop()
            else :
                stack.append(1)
                break
    if len(stack) == 0 :
        print('yes')
    else :
        print('no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;
//파이썬 풀이후 c++풀이
int main(void)
{
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	while (true)
	{
		stack<char> s;
		string inputs;//한줄입력 고생함.
		getline(cin, inputs);
		if (inputs == ".")
		{
			break;
		}
		for (int i = 0; i < inputs.length(); i++)
		{
			if (inputs[i] == '[' || inputs[i] == '(')
			{
				s.push(inputs[i]);
			}
			else if (inputs[i] == ']')
			{
				if (!s.empty() && s.top() == '[')
				{
					s.pop();
				}
				else
				{
					s.push('1');
				}
			}
			else if (inputs[i] == ')')
			{
				if (!s.empty() && s.top() == '(')
				{
					s.pop();
				}
				else
				{
					s.push('1');
				}
			}
		}
		if (s.empty())
		{
			cout << "yes" <<'\n';
		}
		else
		{
			cout << "no" << '\n';
		}
	}
	return 0;
}